### PR TITLE
[WIP] Finetune function and finetune callbacks

### PR DESCRIPTION
**Pull request by TezRomacH**
## What does this PR do?

This PR moves finetune function from `lightning-flash to `pytorch-lightning`.
Discussed in Slack: https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1624628665083500

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tchaton** [comment]
Hey @ananthsub @awaelchli @justusschock, 

Mind sharing your thoughts on this integration of fine-tune within Lightning directly ?

Best,
T.C

File: pytorch_lightning/callbacks/finetuning.py [hunk 1]
```diff
@@ -414,3 +414,147 @@ def finetune_function(self, pl_module: 'pl.LightningModule', epoch: int, optimiz
...
+        self,
+        pl_module: 'pl.LightningModule',
+        epoch: int,
+        optimizer: Optimizer,
+        opt_idx: int,
+    ) -> None:
+        pass
+
+
+class FlashBaseFinetuning(BaseFinetuning):
```
Comment (justusschock) [inline pytorch_lightning/callbacks/finetuning.py:434]: probably not flash, right?

**Review by justusschock — COMMENTED** [review]
I actually find it a bit weird, that something that is "worth" adding a new function to the trainer only has changes in the callbacks.

I think for me, realizing this as callbacks is a bit strange at all. I get why this works and integrates nicely, but from my intuition it should not be implemented as callbacks.
I think it should rather be a kind of new Plugin.

File: pytorch_lightning/callbacks/finetuning.py (hunk 1 above)
Comment (TezRomacH) [reply pytorch_lightning/callbacks/finetuning.py:434]: yeah, that's true. gonna fix it during weekend

**Comment by ananthsub** [comment]
I agree with @justusschock - I understand the appeal from the POV of users, especially Flash users. However, I don't agree with Lightning supporting `finetuning` directly like this for the following reasons:

- Fine-tuning is a specific application of `fit`. It doesn't sit at the same level as `fit`, unlike `validate`/`test`/`predict`. This is seen by `finetune` calling `fit` under the hood. As a result, I don't think this should be a top-level API on the Trainer. 
- Today users can do finetuning in Lightning without calling `trainer.finetune()` - so when should users use `fit` vs `finetune`? the framework offering a dedicated `finetune` entry point encourages users to call this function. In this current implementation, `finetune` enforces that users also must enable fine-tuning callbacks. What happens when these callbacks don't generalize to all use case? Almost certainly, users will end up ignoring the framework's suggestion and call `trainer.fit()` directly, thereby ignoring the framework's suggestion. This is further suboptimal because users are forced to investigate this solution, and subsequently feel they wasted their time chasing a dead end.
- From the framework POV, it's unhealthy if users feel they need to ignore or workaround the framework's suggestions or defaults. It can quickly pervade to other parts of the projects, as users wonder what else they should be skeptical of.

> I think for me, realizing this as callbacks is a bit strange at all. I get why this works and integrates nicely, but from my intuition it should not be implemented as callbacks. I think it should rather be a kind of new Plugin.

+1 to callbacks being strange here. On the docs, callbacks are listed as being for non-essential code. However for fine-tuning here, these callbacks are essential to the training procedure. In general, I'd prefer the callbacks to be as LightningModule-agnostic as possible (ie. they should work on all LightningModules). Right now it seems like Callbacks are a convenient mechanism for codesharing across projects, but I don't think that should be the top criteria. Optimizing for code-resharing at the expense of readability or isolation personally feels like the wrong tradeoff. 

- If core Lightning supported some of these interfaces, we will receive constant feature requests from users for extensions. What criteria do we have for what makes sense for the framework to support vs what users should extend on their own? Pessimistically, this can quickly turn into significant maintenance overhead if core Lightning directly supports a long-tail of features without clear value for all use cases, or users feeling they need to ignore these utilities in favor of their own solutions (again feeling like they're working against the framework vs with the framework). 

Instead, I prefer repos that build on top of Lightning, like Flash, offer these because they can be more selective/opinionated around how they want to offer this functionality based on what type of users or projects they cater to.

**Comment by tchaton** [comment]
> I agree with @justusschock - I understand the appeal from the POV of users, especially Flash users. However, I don't agree with Lightning supporting `finetuning` directly like this for the following reasons:
> 
> * Fine-tuning is a specific application of `fit`. It doesn't sit at the same level as `fit`, unlike `validate`/`test`/`predict`. This is seen by `finetune` calling `fit` under the hood. As a result, I don't think this should be a top-level API on the Trainer.
> * Today users can do finetuning in Lightning without calling `trainer.finetune()` - so when should users use `fit` vs `finetune`? the framework offering a dedicated `finetune` entry point encourages users to call this function. In this current implementation, `finetune` enforces that users also must enable fine-tuning callbacks. What happens when these callbacks don't generalize to all use case? Almost certainly, users will end up ignoring the framework's suggestion and call `trainer.fit()` directly, thereby ignoring the framework's suggestion. This is further suboptimal because users are forced to investigate this solution, and subsequently feel they wasted their time chasing a dead end.
> * From the framework POV, it's unhealthy if users feel they need to ignore or workaround the framework's suggestions or defaults. It can quickly pervade to other parts of the projects, as users wonder what else they should be skeptical of.
> 
> > I think for me, realizing this as callbacks is a bit strange at all. I get why this works and integrates nicely, but from my intuition it should not be implemented as callbacks. I think it should rather be a kind of new Plugin.
> 
> +1 to callbacks being strange here. On the docs, callbacks are listed as being for non-essential code. However for fine-tuning here, these callbacks are essential to the training procedure. In general, I'd prefer the callbacks to be as LightningModule-agnostic as possible (ie. they should work on all LightningModules). Right now it seems like Callbacks are a convenient mechanism for codesharing across projects, but I don't think that should be the top criteria. Optimizing for code-resharing at the expense of readability or isolation personally feels like the wrong tradeoff.
> 
> * If core Lightning supported some of these interfaces, we will receive constant feature requests from users for extensions. What criteria do we have for what makes sense for the framework to support vs what users should extend on their own? Pessimistically, this can quickly turn into significant maintenance overhead if core Lightning directly supports a long-tail of features without clear value for all use cases, or users feeling they need to ignore these utilities in favor of their own solutions (again feeling like they're working against the framework vs with the framework).
> 
> Instead, I prefer repos that build on top of Lightning, like Flash, offer these because they can be more selective/opinionated around how they want to offer this functionality based on what type of users or projects they cater to.

Hey @ananthsub @justusschock, thanks for sharing details thoughts there.

I don't think a Plugin would generalise well enough and would have the same limitation as the current approach.

@ananthsub Your argumentation makes sense and I believe we should integrate features / API design which works for all LightningModule.

@TezRomacH I apologise, but from above conversation, I agree Lightning shouldn't integrate an opinionated solution for fine-tuning as the current approach won't work for a large number of methods even if it covers the most traditional backbone related one.

Best,
T.C

**Comment by TezRomacH** [comment]
Okay!
